SCUMM: Add a tool for extracting lfg archive from version of indy4

phcoder opened the pull request.

phcoder committed:
SCUMM: Add a tool for extracting nd4 archive as used by floppy version of indy4

einstein95 commented:
May want to rename it to something like `extract_lfg`, as the `LFG!` archives were used in more than Indy4, including the German version of MI2 and the LucasArts Classic Adventures

einstein95 commented on engines/scumm/extract_nd4.cpp:
```diff
@@ -0,0 +1,111 @@
+		if (READ_BE_UINT32(head) != MKTAG('L', 'F', 'G', '!')) {
+			fprintf (stderr, "%s doesn't contain magic\n", argv[i]);
+			free (buf);
+			return -2;			
```
trailing whitespace here

einstein95 commented on engines/scumm/extract_nd4.cpp:
```diff
@@ -0,0 +1,111 @@
+			delete[] uncompressedBuf;
+			continue;
+		}
+		
```
same here

einstein95 commented:
Looks like that some files use a dictionary size of 0 which dcl.cpp doesn't support, some games used this for 00.LFL/000.LFL so supporting it is probably needed. Here's a demo that uses it for `READ.ME`
[indy4_demo.zip](https://github.com/scummvm/scummvm-tools/files/10963940/indy4_demo.zip)

dwatteau commented:
Thank you for implementing this!

The format has also been described here for reference: http://seltmannsoftware.blogspot.com/2017/05/adventures-in-useless-reverse.html

I have such an archive for Indy4-dos-fr and one for Monkey4-dos-en. I'll test them with your PR later today.

phcoder commented:
> Looks like that some files use a dictionary size of 0 which dcl.cpp doesn't support, some games used this for 00.LFL/000.LFL so supporting it is probably needed. Here's a demo that uses it for `READ.ME` [indy4_demo.zip](https://github.com/scummvm/scummvm-tools/files/10963940/indy4_demo.zip)

Thank you for this file. It turned out that name field is 13 bytes rather than unlimited null-terminated. After fixing this, it works

phcoder commented:
> May want to rename it to something like `extract_lfg`, as the `LFG!` archives were used in more than Indy4, including the German version of MI2 and the LucasArts Classic Adventures

Thank you. Renamed to extract_lfg and added a bit of more info for useful testing

phcoder commented:
> Thank you for implementing this!
> 
> The format has also been described here for reference: http://seltmannsoftware.blogspot.com/2017/05/adventures-in-useless-reverse.html
> 
> I have such an archive for Indy4-dos-fr and one for Monkey2-dos-en. I'll test them with your PR later today.

Thank you for the link. It cleared up couple of wrong assumptions

dwatteau commented:
FWIW, I may need to wait a few more days to be able to test it with my own set of LFG images; probably this weekend I guess.

dwatteau commented:
Testing from my own LFG archives (which are known to work with the original DOS extractor and with the other extractor that's on Github):

* Indy4, French floppy release put on CD-ROM by Ubisoft: `ATLANTIS.000` is OK, `ATLANTIS.001` has the right number of bytes but a good chunk of them are NUL
* Monkey2, English floppy release from the Limited Run Games set, `MONKEY2.000` is missing, and `MONKEY2.001` is invalid too

@phcoder You can ping me over Discord for easier checking of those. Thanks!

sev- commented:
Merging this, further improvements could be pushed directly.
